### PR TITLE
[Filebeat] [AWS] Add support to source AWS cloudwatch logs from linked accounts

### PR DESCRIPTION
**Pull request by Kavindu-Dodan**
## Proposed commit message

PR adds support to Cloudwatch logs from source linked accounts. This is implemented by using existing configuration `log_group_arn` and mapping it to `LogGroupIdentifier` of FilterLogEvents API [1]

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Note for reviewers
You could review commit by commit for better understanding of the changes 

## How to test this PR locally

This require a linked cloudwatch account. If already has one, then,

- Push logs to a newly created log group OR use an already existing log group in a source account
  - Note - you may use data-gen Go program to generate and push logs to your log group (using  output `CLOUDWATCH_LOG`) [2] 
- Configure filebeat cloudwatch input with log group ARN to `log_group_arn`
- Run filebeat and observe filebeat logs in Kibana discover

## Related issues

Addresses: https://github.com/elastic/beats/issues/36642
And makes https://github.com/elastic/beats/pull/36645 PR obsolete 
closes https://github.com/elastic/beats/issues/37681

## Next step

Utilize `includeLinkedAccounts` when dealing with prefixes. To be done in a dedicated PR.  

[1] - https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html
[1] - https://github.com/Kavindu-Dodan/data-gen

### CODE REVIEW

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -46,6 +46,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - Added `container.image.name` to `journald` Filebeat input's Docker-specific translated fields. {pull}40450[40450]
 - Change log.file.path field in awscloudwatch input to nested object. {pull}41099[41099]
 - Remove deprecated awscloudwatch field from Filebeat. {pull}41089[41089]
+- Add support to source AWS cloudwatch logs from linked accounts. {pull}41188[41188]
```
Comment (kaiyan-sheng) [inline CHANGELOG.next.asciidoc:49]: This is under breaking change section, we can move it to Added section.

File: x-pack/filebeat/docs/inputs/input-aws-cloudwatch.asciidoc [hunk 2]
```diff
@@ -40,16 +40,24 @@ The `aws-cloudwatch` input supports the following configuration options plus the
 [float]
 ==== `log_group_arn`
 ARN of the log group to collect logs from.
+The ARN may refer to a log group in a linked source account.
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/docs/inputs/input-aws-cloudwatch.asciidoc:43]: Can we also make a note here to point out: "If the log group is in a source account and you are using a monitoring account, you must use the log group ARN."

File: CHANGELOG.next.asciidoc (hunk 1 above)
Comment (belimawr) [reply CHANGELOG.next.asciidoc:49]: +1

**Review by belimawr — COMMENTED** [review]
There are just a couple of small things to fix:
- The changelog entry
- Addressing the comment from @kaiyan-sheng.

File: CHANGELOG.next.asciidoc (hunk 1 above)
Comment (Kavindu-Dodan) [reply CHANGELOG.next.asciidoc]: thank you @kaiyan-sheng , I corrected this with https://github.com/elastic/beats/pull/41188/commits/e280d23c0441465e1de27c9dd11e8e2db5eeb11a

File: x-pack/filebeat/docs/inputs/input-aws-cloudwatch.asciidoc (hunk 2 above)
Comment (Kavindu-Dodan) [reply x-pack/filebeat/docs/inputs/input-aws-cloudwatch.asciidoc:43]: Done, please see the commit - https://github.com/elastic/beats/pull/41188/commits/e280d23c0441465e1de27c9dd11e8e2db5eeb11a

**Comment by Kavindu-Dodan** [comment]
@kaiyan-sheng @belimawr thanks for the reviews, I added proposed changes with my latest commit :) appreciate another look

File: x-pack/filebeat/input/awscloudwatch/input.go [hunk 3]
```diff
@@ -120,36 +118,54 @@ func (in *cloudwatchInput) Run(inputContext v2.Context, pipeline beat.Pipeline)
...
-		resourceARNSplit := strings.Split(arnParsed.Resource, ":")
-		if len(resourceARNSplit) >= 2 && resourceARNSplit[0] == "log-group" {
-			return resourceARNSplit[1], arnParsed.Region, nil
+		if parsedArn.Region == "" {
+			return nil, "", fmt.Errorf("failed to parse log group ARN: missing region")
 		}
+
+		// refine to match AWS API parameter regex of logGroupIdentifier
+		groupId := strings.TrimSuffix(cfg.LogGroupARN, ":*")
+		logGrouIDs = append(logGrouIDs, groupId)
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awscloudwatch/input.go:147]:
```suggestion
		logGroupIDs = append(logGroupIDs, groupId)
```

File: x-pack/filebeat/input/awscloudwatch/cloudwatch.go [hunk 4]
```diff
@@ -96,16 +96,16 @@ func (p *cloudwatchPoller) getLogEventsFromCloudWatch(svc *cloudwatchlogs.Client
...
 	return nil
 }
 
-func (p *cloudwatchPoller) constructFilterLogEventsInput(startTime, endTime time.Time, logGroup string) *cloudwatchlogs.FilterLogEventsInput {
+func (p *cloudwatchPoller) constructFilterLogEventsInput(startTime, endTime time.Time, logGroupId string) *cloudwatchlogs.FilterLogEventsInput {
 	filterLogEventsInput := &cloudwatchlogs.FilterLogEventsInput{
-		LogGroupName: awssdk.String(logGroup),
-		StartTime:    awssdk.Int64(unixMsFromTime(startTime)),
-		EndTime:      awssdk.Int64(unixMsFromTime(endTime)),
+		LogGroupIdentifier: awssdk.String(logGroupId),
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awscloudwatch/cloudwatch.go:106]: Do we need to check if logGroupId is ARN, then we use `LogGroupIdentifier` in the input, if logGroupId is name, we use `logGroupName` instead?

File: x-pack/filebeat/input/awscloudwatch/input.go (hunk 3 above)
Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/awscloudwatch/input.go]: fixed in the latest commit - https://github.com/elastic/beats/pull/41188/commits/e8f2ef9fbe3cc8e1007172d07c594d4b97f7fcec

File: x-pack/filebeat/input/awscloudwatch/cloudwatch.go (hunk 4 above)
Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/awscloudwatch/cloudwatch.go:106]: `logGroupIdentifier` accepts both ARN & name (see field description at [1]). So we can use input from `log_group_arn` &  `log_group_name` as it is (as long as they are valid). 

[1] - https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html

File: x-pack/filebeat/input/awscloudwatch/input.go [hunk 5]
```diff
@@ -120,36 +118,54 @@ func (in *cloudwatchInput) Run(inputContext v2.Context, pipeline beat.Pipeline)
...
 	}
 
+	return logGroupIDs, region, nil
+}
+
+// getLogGroupNames uses DescribeLogGroups API to retrieve all log group names
+func getLogGroupNames(svc *cloudwatchlogs.Client, logGroupNamePrefix string) ([]string, error) {
 	// construct DescribeLogGroupsInput
 	describeLogGroupsInput := &cloudwatchlogs.DescribeLogGroupsInput{
 		LogGroupNamePrefix: awssdk.String(logGroupNamePrefix),
```
Comment (stefanhettich) [inline x-pack/filebeat/input/awscloudwatch/input.go:171]: Might the attribute `IncludeLinkedAccounts: awssdk.Bool(true)` be missing here? For me, it seems to be required to get the log groups from (all) linked accounts. Correct me, if I'm wrong!

Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/awscloudwatch/input.go:171]: @stefanhettich really good point and thank you bringing this to my attention :) 

Yes, per the API DescribeLogGroups [1] `includeLinkedAccounts` needs to be true to include linked accounts.

Further, this opens up another input possibility. If we utilize `includeLinkedAccounts = true` & use returned `LogGroupArn`, then users can utilize `logGroupNamePrefix`and still use linked accounts. 

I will validate this in my setup and update accordingly :) 


[1] https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_DescribeLogGroups.html

Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/awscloudwatch/input.go:171]: @kaiyan-sheng fyi, this is something we didn't consider :)

Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/awscloudwatch/input.go:171]: @stefanhettich  again, this was a great input and I am considering adding support to this. But I will add this through a dedicated PR and make sure both are included in the next release (aiming for 8.15.4).

Also please check the original issue #36642 and comment [1] to track updates 

@kaiyan-sheng  will open a follow-up as discussed :) 

[1] - https://github.com/elastic/beats/issues/36642#issuecomment-2406110947

**Review by kaiyan-sheng — APPROVED** [review]

**Comment by Kavindu-Dodan** [comment]
@belimawr appreciate another review from you :)

File: x-pack/filebeat/input/awscloudwatch/input.go (hunk 5 above)
Comment (stefanhettich) [reply x-pack/filebeat/input/awscloudwatch/input.go:171]: Awesome! Thanks a lot for taking care. Feel free to reach out, if I can help with anything! :)

Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/awscloudwatch/input.go:171]: Opened dedicated PR https://github.com/elastic/beats/pull/41206 to add support to this :) Will resolve this conversation now

**Review by belimawr — APPROVED** [review]

**Pull request merged**